### PR TITLE
Variant utilities

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: *(direct link to the issue)*

Variant::VariantVisitor is a useful utility template for defining visitors of variants, uniting any
number of callables into one set of overload resolution candidates.

Variant::Visit() now simulates the one-visitor case of C++17 std::visit more exactly according to the standard.

This includes perfect forwarding of return values, the requirement that all alternatives return exactly the same type,
allowing the variant argument to be a proper subclass of a std::variant specialization, and sfinae to exclude other
calls to Visit from overload resolution.

Also, the performance guarantee of the standard, that dispatch is constant-time.

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Review by crsib — APPROVED** [review]
Pretty awesome!

**Review by crsib — CHANGES_REQUESTED** [review]
I'll like the proposed changes, but GCC does not :-)

**Comment by Paul-Licameli** [comment]
> I'll like the proposed changes, but GCC does not :-)

It may be GCC isn't doing CTAD correctly and a `make_visitor` function like `make_pair` may be needed after all

**Comment by Paul-Licameli** [comment]
This idea needed much more work than I thought to get all the details really right.

Please review the whole again from the beginning.

There are also now many more compile-time tests (so we know the cases don't violate the `static_assert`s) and run-time tests of various combinations of values in the visitor, and constness and reference types of the arguments and the expected return value type.

File: libraries/lib-utility/Variant.h [hunk 1]
```diff
@@ -31,57 +33,76 @@ struct VisitHelperReturn {
...
+   assert(pValue);
+   if constexpr (std::is_lvalue_reference_v<Variant>)
+      return std::invoke(std::forward<Visitor>(vis), (*pValue));
+   else
+      return std::invoke(std::forward<Visitor>(vis), std::move(*pValue));
+}
+
+//! Help to define Visit() below
+template <size_t Index, typename Visitor, typename Variant>
+auto TypeCheckedVisitHelperFunction(Visitor &&vis, Variant &&var)
```
Comment (crsib) [inline libraries/lib-utility/Variant.h:67]: I wonder if `decltype(auto)` is needed here as well

Comment (crsib) [reply libraries/lib-utility/Variant.h:67]: Or, sorry, I have missed the `->` arrow

File: libraries/lib-utility/Variant.h [hunk 2]
```diff
@@ -101,18 +132,18 @@ VisitHelper(std::index_sequence<Indices...>, Visitor &&vis, Variant &&var)
  so let's use this even when not needed on the other platforms, instead of
  having too much conditional compilation
  */
-template <typename Visitor, typename Variant>
+template <typename Visitor, typename Variant,
+   typename VariantBase = detail::deduced_variant<Variant &&>>
```
Comment (crsib) [inline libraries/lib-utility/Variant.h:136]: Pretty cool!

**Review by crsib — APPROVED** [review]
Looks fantastic!

**Pull request merged**